pkg/controller/factory/base_controller.go

```diff
@@ -62,19 +63,33 @@ func (s *scheduledJob) Run() {
 	s.queue.Add(DefaultQueueKey)
 }
 
+func waitForNamedCacheSync(controllerName string, stopCh <-chan struct{}, cacheSyncs ...cache.InformerSynced) error {
+	klog.Infof("Waiting for caches to sync for %s", controllerName)
+
+	if !cache.WaitForCacheSync(stopCh, cacheSyncs...) {
+		return fmt.Errorf("unable to sync caches for %s", controllerName)
+	}
+
+	klog.Infof("Caches are synced for %s ", controllerName)
+
+	return nil
+}
+
 func (c *baseController) Run(ctx context.Context, workers int) {
 	// HandleCrash recovers panics
 	defer utilruntime.HandleCrash()
 
 	// give caches 10 minutes to sync
 	cacheSyncCtx, cacheSyncCancel := context.WithTimeout(ctx, c.cacheSyncTimeout)
 	defer cacheSyncCancel()
-	if !cache.WaitForNamedCacheSync(c.name, cacheSyncCtx.Done(), c.cachesToSync...) {
-		// the parent context is closed, it means we are shutting down, do not call panic()
-		if ctx.Err() != nil {
+	err := waitForNamedCacheSync(c.name, cacheSyncCtx.Done(), c.cachesToSync...)
+	if err != nil {
+		select {
+		case <-ctx.Done():
+			// Exit gracefully
 			return
-		}
-		panic("failed to wait for cache to sync")
+		default:
+			klog.Exit(err)
```

Comment:
add a comment explaining
```
// if caches did not sync after 10 minutes, it has taken oddly long and we should provide feedback. Since the control loops will never start, it is safer to exit with a good message than to continue with a dead loop. 
// TODO consider making this behavior configurable.
```

Reply:
done